rust_src/src/composite.rs

```diff
@@ -0,0 +1,79 @@
+//! Composite sequence support.
+
+use remacs_macros::lisp_fn;
+
+use crate::{
+    fns::validate_subarray_rust,
+    lisp::LispObject,
+    multibyte::LispStringRef,
+    remacs_sys::{EmacsInt, Fput_text_property, Qcomposition},
+};
+
+/// Internal use only.
+///
+/// Compose text between indices START and END of STRING, where
+/// START and END are treated as in `substring'.  Optional 4th
+/// and 5th arguments are COMPONENTS and MODIFICATION-FUNC
+/// for the composition.  See `compose-string' for more details.
+#[lisp_fn(min = "3")]
+pub fn compose_string_internal(
+    string: LispStringRef,
+    start: EmacsInt,
+    end: EmacsInt,
+    components: LispObject,
+    modification_func: LispObject,
+) -> LispObject {
+    let (from, to) =
+        validate_subarray_rust(string.into(), start.into(), end.into(), string.len_chars());
+
+    compose_text_rust(from, to, components, modification_func, string);
+
+    LispObject::from(string)
```

Comment:
You can simply return the string and make the return type `LispStringRef`. The `lisp_fn` macro handles the types coming and and going out.